### PR TITLE
HPCC-28507 Display file min and max skews as percentages

### PR DESCRIPTION
**Pull request by jeclrsg**
Signed-off-by: Jeremy Clements <Jeremy.Clements@lexisnexisrisk.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [x] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: esp/src/eclwatch/LFDetailsWidget.js [hunk 1]
```diff
@@ -462,6 +464,12 @@ define([
                 this.updateInput("RecordSize", oldValue, this.i18n.NoPublishedSize);
             } else if (name === "Cost") {
                 this.updateInput("FormattedCost", oldValue, Session.formatCost(newValue));
+            } else if (name === "MinSkew") {
+                var formatNum = hpccCommon.format(",.2f");
+                domAttr.set(this.id + "MinSkew", "innerHTML", this.logicalFile.MinSkew ? formatNum(this.logicalFile.MinSkew) + "%" : "");
+            } else if (name === "MaxSkew") {
+                var formatNum = hpccCommon.format(",.2f");
```
Comment (GordonSmith) [inline esp/src/eclwatch/LFDetailsWidget.js:471]: formatNum declaration should be moved to module scope or the Util module as "skewFormat"?

File: esp/src/eclwatch/DFUQueryWidget.js [hunk 2]
```diff
@@ -649,6 +651,7 @@ define([
 
         initWorkunitsGrid: function () {
             var context = this;
+            var formatNum = hpccCommon.format(",.2f");
```
Comment (GordonSmith) [inline esp/src/eclwatch/DFUQueryWidget.js:654]: formatNum declaration should be moved to module scope or the Util module as "skewFormat"?

File: esp/src/src/Session.ts [hunk 3]
```diff
@@ -48,7 +48,8 @@ getBuildInfo().then(info => {
     dojoConfig.currencyCode = info["currencyCode"] ?? "";
 });
 
-const format = d3Format(".2f");
+const format = d3Format(",.2f");
+export const formatNum = format;
```
Comment (GordonSmith) [inline esp/src/src/Session.ts:52]: Could be in a module that is accessible from Dojo and React?

**Review by GordonSmith — CHANGES_REQUESTED** [review]

File: esp/src/src/Utility.ts [hunk 4]
```diff
@@ -1061,6 +1061,14 @@ export function deleteCookie(name: string) {
     document.cookie = `${name}=; domain=${window.location.hostname}; path=/; expires=${expireDate.toUTCString()}`;
 }
 
+export function formatDecimal(str): string {
+    const format = d3Format(",.2f");
```
Comment (GordonSmith) [inline esp/src/src/Utility.ts:1065]: This should be moved into module scope

**Review by GordonSmith — CHANGES_REQUESTED** [review]

File: esp/src/eclwatch/DFUQueryWidget.js [hunk 5]
```diff
@@ -736,13 +736,13 @@ define([
                     MinSkew: {
                         label: nlsHPCC.MinSkew, width: 60,
                         formatter: function (value, row) {
-                            return value !== undefined ? value : "";
+                            return value ? "-" + Utility.formatDecimal(value / 100) + "%" : "";
```
Comment (GordonSmith) [inline esp/src/eclwatch/DFUQueryWidget.js:739]: I don't think we actually display the "-" anywhere else for skews?

File: esp/src/src-react/components/Files.tsx [hunk 6]
```diff
@@ -200,10 +200,10 @@ export const Files: React.FunctionComponent<FilesProps> = ({
                 }, []),
             },
             MinSkew: {
-                label: nlsHPCC.MinSkew, width: 60, formatter: React.useCallback((value, row) => value ?? "", [])
+                label: nlsHPCC.MinSkew, width: 60, formatter: React.useCallback((value, row) => value ? `-${Utility.formatDecimal(value / 100)}%` : "", [])
```
Comment (GordonSmith) [inline esp/src/src-react/components/Files.tsx:203]: I don't think we want the "-"

**Review by GordonSmith — CHANGES_REQUESTED** [review]

File: esp/src/eclwatch/DFUQueryWidget.js [hunk 7]
```diff
@@ -736,13 +736,21 @@ define([
                     MinSkew: {
                         label: nlsHPCC.MinSkew, width: 60,
                         formatter: function (value, row) {
-                            return value !== undefined ? value : "";
+                            if (value.toString().indexOf(".") >= 0) {
```
Comment (GordonSmith) [inline esp/src/eclwatch/DFUQueryWidget.js:739]: value could be undefined?

**Review by GordonSmith — CHANGES_REQUESTED** [review]

File: esp/src/src-react/components/Files.tsx [hunk 8]
```diff
@@ -200,10 +200,10 @@ export const Files: React.FunctionComponent<FilesProps> = ({
                 }, []),
             },
             MinSkew: {
-                label: nlsHPCC.MinSkew, width: 60, formatter: React.useCallback((value, row) => value ?? "", [])
+                label: nlsHPCC.MinSkew, width: 60, formatter: React.useCallback((value, row) => value ? `${Utility.formatDecimal(value / 100)}%` : "", [])
```
[the remark this answers is not in the document]
Comment (jeclrsg) [reply esp/src/src-react/components/Files.tsx:203]: There's that ternary there that should be handling that, returning ""

File: esp/src/eclwatch/DFUQueryWidget.js [hunk 9]
```diff
@@ -736,13 +736,27 @@ define([
                     MinSkew: {
                         label: nlsHPCC.MinSkew, width: 60,
                         formatter: function (value, row) {
-                            return value !== undefined ? value : "";
+                            if (value) {
```
Comment (GordonSmith) [inline esp/src/eclwatch/DFUQueryWidget.js:739]: Value could equal `0`

File: esp/src/eclwatch/DFUQueryWidget.js [hunk 10]
```diff
@@ -736,13 +736,27 @@ define([
...
+                                }
+                            }
+                            return "";
                         }
                     },
                     MaxSkew: {
                         label: nlsHPCC.MaxSkew, width: 60,
                         formatter: function (value, row) {
-                            return value !== undefined ? value : "";
+                            if (value) {
```
Comment (GordonSmith) [inline esp/src/eclwatch/DFUQueryWidget.js:752]: Value could equal `0`

**Review by GordonSmith — CHANGES_REQUESTED** [review]

**Review by GordonSmith — APPROVED** [review]

**Comment by jeclrsg** [comment]
rebased to fix a merge conflict

**Comment by GordonSmith** [comment]
@ghalliday this is ready for pull

**Pull request merged**